src/materials/ConcreteASREigenstrain.C

```diff
@@ -0,0 +1,323 @@
+    _console << iter_output.str();
+  if (it == _max_its && norm_residual > _absolute_tolerance &&
+      (norm_residual / first_norm_residual) > _relative_tolerance)
+  {
+    if (_output_iteration_info_on_error)
+      Moose::err << iter_output.str();
+    mooseError("Max material iteration hit during nonlinear constitutive model solve!");
+  }
+
+  // new ASR reaction extent
+  _ASR_extent[_qp] = scalar;
+  Real inc_ASR_extent = _ASR_extent[_qp] - _ASR_extent_old[_qp];
+
+  // stress dependent total ASR volumetric accounting for ASR gel absorption due to tensile and
+  // compressive cracking Eigen solve - Note the eigen values are ranked from minimum to maximum
+  const Real sig_k = _stress_eigenvalues[2];
+
+  Real gel_absorption_tensile = 1.0;
+  Real gel_absorption_compress = 1.0;
+
+  // Optionally calculate effect of ASR reaction on the tensile strength
+  Real f_t = _f_tensile;
+  if (_ASR_dependent_tensile_strength)
+    f_t = _f_tensile * (1.0 - (1.0 - _beta_f) * _ASR_extent[_qp]);
+
+  if (sig_k > _gamma_tensile * f_t)
+    gel_absorption_tensile =
+        _gamma_residual + (1.0 - _gamma_residual) * (_gamma_tensile * f_t / sig_k);
+
+  Real sig_effective = _stress[_qp].trace() / (3.0 * -_f_compress);
+
+  if (sig_effective > 0.0)
+  {
+    gel_absorption_compress =
+        1.0 - std::exp(_beta) * sig_effective / (1.0 + (std::exp(_beta) - 1.0) * sig_effective);
+    if (gel_absorption_compress > 1.0)
+      gel_absorption_compress = 1.0;
+    if (gel_absorption_compress < 0.0)
+      gel_absorption_compress = 0.0;
+  }
+
+  const Real inc_ASR_volumetric_strain = gel_absorption_tensile * gel_absorption_compress *
+                                         std::pow(_rh[_qp], _rh_exponent) * inc_ASR_extent *
+                                         _max_vol_expansion;
+
+  return _volumetric_strain_old[_qp] + inc_ASR_volumetric_strain;
+}
+
+Real
+ConcreteASREigenstrain::computeResidual(unsigned qp, Real scalar)
+{
+  Real f;
+  if (_expansion_type == ExpansionType::Isotropic)
+    f = 1.0;
+  else if (_expansion_type == ExpansionType::Anisotropic)
+  {
+    const Real I_sigma = _stress[_qp].trace();
+    if (I_sigma >= 0.0) // hydrostatic tension
+      f = 1.0;
+    else // hydrostatic compression: retarding ASR rection
+    {
+      f = 1.0 + _alpha * I_sigma / (3.0 * -_f_compress);
+      if (f < 1.0)
+        mooseError("Wrong retardation for ASR latency time calculation!");
+    }
+  }
+  else
+    mooseError("Invalid expansion type");
+
+  // Convert current temperature to Kelvin
+  const Real T = _temperature[qp] + _temp_offset;
+
+  // ASR characteristic and latency times (in days)
+  Real tau_c = _tau_c_T0 * std::exp(_Uc * (1.0 / T - 1.0 / _ref_temp));
+  Real tau_L = f * _tau_L_T0 * std::exp(_UL * (1.0 / T - 1.0 / _ref_temp));
+
+  Real resid = scalar - _ASR_extent_old[qp] -
+               _dt / 86400.0 / tau_c * (1.0 - scalar) / (1.0 + std::exp(-tau_L / tau_c)) *
+                   (scalar + std::exp(-tau_L / tau_c));
```

Comment:
Wow this is complicated. Would you separate out into a few different variables or add in some parenthesis to make the equation more human readable?

Reply:
OK, I have an improvement.